### ISSUE TITLE
0.4 issues tracking

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#5673]
#### Core checklist

- [x] `torch.sum(dtype=)` (make it take an optional dtype and have a sensible default dtype for integer stuff; e.g., to avoid overflow) (PR: https://github.com/pytorch/pytorch/pull/6573)
- [x] Bind integer scalars as `int64_t` in PyTorch-ATen conversion, also allow integer scalars as members in IntList? (https://github.com/pytorch/pytorch/pull/6637)
- [x] Make concrete type of Tensor `torch.Tensor`, not `torch.autograd.Variable`
- [x] Fix printing, esp. of scalars
- [x] Factory function changes @gchanan (e.g., how to create sparse tensors) (https://github.com/pytorch/pytorch/issues/5641)
- [x] ~~Scalar issues (scalar + tensor doesn't work, Mixed precision binary ops for scalars) #5647~~
- [x] intra-dtype conversions; i.e. torch.int64.to_sparse() should return torch.sparse.int64).
- [x] Greater than 2 node distributed training is broken @ailzhang 
- [x] DataParallel converges to worse accuracy than 0.3 (?) @teng-li 

#### Deferred until later

- ~~Performance regression when indexing #5388~~

#### Binaries checklist:

- [x] Fix pytorch.version linker script to work for master
- [x] add a build flag for statically linking cublas, cusparse, cudnn, nccl2
- [x] In docker image, strip unsupported architecture symbols in libcublas_static.a, libcusparse_static.a, libcudnn_static.a, libnccl_static.a using `nvprune`
  - Supported: 3.5, 5.2, 6.0, 6.1, 7.0
- [x] fix "invalid pointer" on error messages https://github.com/pytorch/pytorch/issues/5400
- [x] ~~(maybe) build static fPIC openmp and link against that instead (requires recompiling gcc from source)~~
- [x] use devtoolset3 (for gcc 4.9.2+, because cpp_extensions need that)

Regressions checklist:
- [x] test final binaries against community repos and do basic smoke tests
- [x] check basic perf-regressions against examples/

#### Thirdparty-repo checklist:

- [x] Update examples repositories
- [x] update vision
- [ ] update audio (merge https://github.com/pytorch/audio/pull/34 and remerge https://github.com/pytorch/audio/pull/23 )

#### Documentation checklist:

- [x] write migration guide
- [x] write release notes / blogpost
- [x] take a pass on tutorials
- [x] Merge https://github.com/pytorch/tutorials/pull/214 and mention it in the release notes
- [x] Documentation rewrite, especially of autograd

#### Windows checklist:

- [x] build final conda binaries
- [x] build wheels (@peterjc123)
- [x] fix dataloader memory leak https://github.com/pytorch/pytorch/pull/5585
- [ ] check off https://github.com/pytorch/pytorch/issues/4092

**Comment by apaszke** [comment]
We also need to fix the scalar issues, see e.g. #5647

**Comment by SsnL** [comment]
We should also fix indexing perf regression, right? https://github.com/pytorch/pytorch/issues/5388

**Comment by gchanan** [comment]
@ezyang does dtype conversions mean conversions _within_ dtypes, i.e. I should be able to do torch.int64.to_sparse() and get torch.sparse.int64?

**Comment by ezyang** [comment]
@gchanan Either you or @colesbury stated this one. I think the intent was any missing dtype conversions that people would need.

**Comment by peterjc123** [comment]
@ezyang What part of autograd doc has to be rewritten? I don't quite get your idea. Could you give me an example or sth like that? And does the second point imply that we don't need to build wheel packages for Windows?

**Comment by fmassa** [comment]
There are a few performance regressions in indexing that I think are not mentioned here but would be good to keep in mind for before the release:
- [ ] indexing a single dimension in a multi-dimensional tensor (which was before `index_select`), see  https://github.com/pytorch/pytorch/issues/5159#issuecomment-364627840 . 
- [ ] indexing with a byte tensor of the same size of the tensor (which was before delegated to `masked_select`), see https://github.com/pytorch/pytorch/pull/3725#issuecomment-345059385

**Comment by ezyang** [comment]
@peterjc123 It's miscategorized; I moved it. Wheels for Windows sounds useful; can I put you in charge of those?

**Comment by peterjc123** [comment]
@ezyang Sure.

**Comment by zou3519** [comment]
I think we have issues around binary ops with LongTensors and floats (like #5694 and #5394), and binary ops between LongTensors (like #5411) where the behavior is unintuitive for people coming from Python and numpy. It would be nice to fix these in general by the release but I'm not sure how feasible that is.

**Comment by SsnL** [comment]
Can we update this please?
